[Tools] Create swift-inspect, a debugging tool for dumping Swift runtime data from another process

mikeash opened the pull request:
Create `swift-inspect`. This is a tool which can inspect a Swift process and dump information about the Swift runtime in that process.

It currently supports inspecting two kinds of information:

1. It can dump the protocol conformance cache.
2. It can dump all metadata allocations, and print the name and size of the metadata allocated by the generic metadata cache.

It's meant to grow more functionality over time.

The design of the tool places all of the runtime-related smarts in Remote Mirror. `swiftdt` is then a small that connects the Remote Mirror functionality with remote process inspection. The remote process inspection parts are necessarily OS-specific. Currently they're only implemented for Apple OSes. Everything is cleanly separated out so that someone who knows the appropriate APIs for another OS should be able to add support for that OS without much difficulty.

Update: after much soul searching, we have renamed this from `swiftdt` to `swift-inspect`, which is more descriptive and better fits the naming conventions of other tools.

mikeash committed:
[Tools] Start implementing swifdt, a tool to inspect the Swift runtime in another process.

rdar://problem/55481578

mikeash committed:
[Tools] Get swiftdt calling into Symbolication and have it print all symbols from libswiftCore.

rdar://problem/55481578

mikeash committed:
[Tools] Start hooking up swiftdt to Symbolication, print out found symbols and read the contents of Conformances.

rdar://problem/55481578

mikeash committed:
[Tools] Start hooking up swiftdt to Remote Mirror.

rdar://problem/55481578

mikeash committed:
[Tools] Fill out Remote Mirror callbacks for swiftdt.

rdar://problem/55481578

mikeash committed:
[Tools] Get a first pass of swiftdt conformance cache dumping up and running.

rdar://problem/55481578

mikeash committed:
[Tools] Teach swiftdt how to retrieve ObjC class names.

rdar://problem/55481578

mikeash committed:
[Tools] Super rough draft of swiftdt dumping MetadataAllocator contents.

rdar://problem/55481578

mikeash committed:
[Tools] Add rudimentary command parsing to swiftdt.

rdar://problem/55481578

mikeash committed:
[Tools] Add some metadata printing to swiftdt.

rdar://problem/55481578

mikeash committed:
[Runtime] Remove inactive code that looked up entries in the protocol conformance cache by type descriptor

mikeash committed:
[Tools] Simplify swiftdt conformance cache dumping since the conformance cache keys are always metadata pointers.

mikeash committed:
[Tools] Move metadata dumping into swiftdt, have Remote Mirror just iterate and inspect.

rdar://problem/55481578

mikeash committed:
[Reflection] Add error messages to the new Remote Mirror calls and plumb them through swiftdt.

rdar://problem/55481578

mikeash committed:
[Runtime] Only enable metadata allocation tracking when requested with an environment variable.

Expose _swift_debug variables for Remote Mirror to use for tracking metadata allocation, and adopt them in Remote Mirror.

rdar://problem/55481578

mikeash committed:
[Runtime] Expose the protocol conformance state as a _swift_debug variable.

Have Remote Mirror look that up instead of the C++ mangled name.

rdar://problem/55481578

mikeash committed:
[RemoteMirror] Add comments to the new APIs and types.

rdar://problem/55481578

mikeash committed:
[Reflection] Clean up the ReflectionContext additions.

Move the internal runtime structures into RuntimeInternals.h. Remove leftover methods. Comment the new methods.

rdar://problem/55481578

mikeash committed:
[Tools] Add a reminder to run the tool as root if the target can't be inspected.

rdar://problem/55481578

mikeash committed:
[Tools] Get swiftdt actually building again.

rdar://problem/55481578

mikeash committed:
Minor cleanup of initial swiftdt work.

mikeash committed:
[Runtime] Clean up metadata allocation tracking.

Keep the initial metadata pool to be all zeroes. Remove the hardcoded trailer at the end. Write a trailer into it when we check the environment variable the first time we allocate.

gottesmm commented:
@mikeash what does dt stand for? Maybe it would make sense to expand out the name? My thought: swift_memory_dumping_tool?

gottesmm commented on include/swift/Demangling/ManglingUtils.h:
```diff
@@ -191,8 +191,8 @@ void mangleIdentifier(Mangler &M, StringRef ident) {
       M.Buffer << (Repl.StringPos - Pos);
-      assert(!isDigit(ident[Pos]) &&
-             "first char of sub-string may not be a digit");
+//       assert(!isDigit(ident[Pos]) &&
```
Missed this.

mikeash replied:
Thanks. I'm fixing this up along with some other stuff. Not quite done with it yet!

mikeash commented:
@gottesmm It stands for Swift Debug Tool. There's some internal precedent with other `blahdt` tools that it's named after, but I'm not completely opposed to changing it.

gottesmm commented:
@mikeash does it debug everything about a swift program? To me debug tool is very generic to the point of being confusing = /.

MadCoder commented:
it stands for debug/dump tool

mikeash committed:
[Tools] Remove Foundation dependency from swiftdt, which gets very complicated for building.

Add a shim header and corresponding module to stub out a couple of C types we need.

gparker42 commented:
"swiftrtdt" for Swift runtime debug tool?

gparker42 commented on tools/swiftdt/main.swift:
```diff
@@ -0,0 +1,166 @@
+}
+
+if let commandName = commandName {
+  print("error: \(executableName): unknown command \(commandName)", to: &Std.err)
```
Can these both be argFail() calls?

gparker42 commented on tools/swiftdt/main.swift:
```diff
@@ -0,0 +1,166 @@
+}
+
+let commandName = argv.popFirst()
+for command in commands {
```
Check that argv is now empty here, or else allow multiple commands to be run in one execution.

gottesmm commented:
@MadCoder I figured as much. I was just pointing out the ambiguity of the name.

gottesmm commented:
@gparker42 I think swift runtime debug tool would be a better name. At least it is clear what you are debugging.

gottesmm commented on tools/swiftdt/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+add_swift_target_executable(swiftdt
```
This is the wrong way to do this. We do not want add_swift_target_executable to be used outside of the stdlib directory. This is in preparation for splitting the stdlib from the main host build.

If you need to be in swift's cmake build, I would suggest putting this tool into ./stdlib/tools.

Otherwise, if you can compile against a just built toolchain, I would create a build-script tool that builds against the just built toolchain.

gottesmm replied:
I imagine that you can do the latter since you are just using remote mirrors.

gparker42 commented on tools/swiftdt/main.swift:
```diff
@@ -0,0 +1,166 @@
+}
+
+func printUsage(args: inout ArraySlice<String>) {
+  print("Usage: \(executableName) <command>", to: &Std.err)
```
`print("Usage: \(executableName) <pid | process name> <command>", to: &Std.err)`

mikeash replied:
@gottesmm helped me transform this into a proper package, so I went ahead and redid it to use `ArgumentParser` instead of rolling my own. Now we get all the right behavior for free!

gottesmm commented:
@mikeash one cool thing is that if you make this a build-script tool, you can use the swift package manager and get access to the argument parser/other things.

gottesmm commented:
*use the just built swiftpm with your project.

gottesmm committed:
[build-script] Convert swiftdt to be a swiftpm project that is built via a build-script product.

gottesmm commented:
@mikeash #31667. I tested that locally. It works. To build, one does:

#31667

If one wants to invoke the build_script_helper.py directly, one does it like this:

```
+./build_script_helper.py --toolchain $PWD/../../../../build/Ninja-RelWithDebInfoAssert/toolchain-macosx-x86_64/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain --build-path $PWD/build --package-path $PWD
```

Or to do it using build-script normally:

```
./swift/utils/build-script --release-debuginfo --swiftdt --skip-build-benchmarks --force-optimized-typechecker --swiftpm --llbuild   --install-swift --install-swiftpm --install-llbuild
```

mikeash committed:
[swiftdt] Switch to ArgumentParser instead of handcrafted argument parsing

rdar://problem/55481578

mikeash committed:
[swiftdt] Import Foundation again, and get rid of our horrible handcrafted NSString bridging.

rdar://problem/55481578

mikeash committed:
[Reflection] Remove useless TaggedUnion.h include. Comment swift_metadata_allocation_tag.

mikeash committed:
[Runtime] Comment the MetadataAllocatorTags enum.

mikeash committed:
[CMake] Undo some changes made for swiftdt.

airspeedswift committed:
nfc: std lib style tweaks

benrimmington commented:
Should this tool be renamed? Other tools in `*.xctoolchain/usr/bin/` include:

```
swift-build
swift-build-tool
swift-demangle
swift-package
swift-run
swift-stdlib-tool
swift-test
```

With this naming scheme, they can be launched as subcommands.
(e.g. `swift stdlib-tool --help`)

airspeedswift committed:
Wrap allocations/metadatas

natecook1000 commented on tools/swiftdt/Sources/swiftdt/main.swift:
```diff
@@ -0,0 +1,122 @@
+  nameOrPid: String
+) -> (Inspector, SwiftReflectionContextRef) {
+  guard let pid = pidFromHint(nameOrPid) else {
+    argFail("Cannot find pid/process \(nameOrPid)")
```
If you don’t mind making this a throwing method, you can use `ValidationError` here.
```suggestion
    throw ValidationError("Cannot find pid/process \(nameOrPid)")
```

airspeedswift committed:
Rename swiftdt to swift-inspect

airspeedswift committed:
Add copywrite headers

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test osx platform

mikeash commented:
@swift-ci please test os x platform

mikeash commented:
@swift-ci please test os x platform

airspeedswift committed:
Separate raw metadata and generic metadata dump commands

airspeedswift commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash committed:
Add swift-inspect default values to expected_options.py.

mikeash commented:
@swift-ci please test

mikeash committed:
Fix python lint errors.

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test Linux platform

airspeedswift reviewed (approved):
:shipit:
